File: src/apps/kernel/arch/gem5/KDTU.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/Common.h>
#include <m3/util/Sync.h>
#include <m3/DTU.h>

#include "../../MainMemory.h"
#include "../../KDTU.h"
#include "../../KVPE.h"

namespace m3 {

void KDTU::do_set_vpeid(size_t core, int oldVPE, int newVPE) {
    alignas(DTU_PKG_SIZE) DTU::reg_t vpeId = newVPE;
    Sync::compiler_barrier();
    write_mem_at(core, oldVPE,
        DTU::dtu_reg_addr(DTU::DtuRegs::VPE_ID), &vpeId, sizeof(vpeId));
}

void KDTU::do_ext_cmd(KVPE &vpe, DTU::reg_t cmd) {
    alignas(DTU_PKG_SIZE) DTU::reg_t reg = cmd;
    Sync::compiler_barrier();
    write_mem(vpe, DTU::dtu_reg_addr(DTU::DtuRegs::EXT_CMD), &reg, sizeof(reg));
}

void KDTU::clear_pt(uintptr_t pt) {
    // clear the pagetable
    char buffer[1024];
    memset(buffer, 0, sizeof(buffer));
    uintptr_t addr = DTU::noc_to_virt(pt);
    for(size_t i = 0; i < PAGE_SIZE / sizeof(buffer); ++i)
        write_mem_at(MEMORY_CORE, 0, addr + i * sizeof(buffer), buffer, sizeof(buffer));
}

void KDTU::init() {
    do_set_vpeid(KERNEL_CORE, KVPE::INVALID_ID, 0);
}

void KDTU::deprivilege(int core) {
    // unset the privileged flag
    alignas(DTU_PKG_SIZE) DTU::reg_t status = 0;
    Sync::compiler_barrier();
    write_mem_at(core, KVPE::INVALID_ID,
        DTU::dtu_reg_addr(DTU::DtuRegs::STATUS), &status, sizeof(status));
}

void KDTU::set_vpeid(int core, int vpe) {
    // currently, the invalid ID is still set, so specify that
    do_set_vpeid(core, KVPE::INVALID_ID, vpe);
}

void KDTU::unset_vpeid(int core, int vpe) {
    do_set_vpeid(core, vpe, KVPE::INVALID_ID);
}

void KDTU::wakeup(KVPE &vpe) {
    // write the core id to the PE
    alignas(DTU_PKG_SIZE) CoreConf conf;
    conf.coreid = vpe.core();
    Sync::compiler_barrier();
    write_mem(vpe, CONF_GLOBAL, &conf, sizeof(conf));

    do_ext_cmd(vpe, static_cast<DTU::reg_t>(DTU::ExtCmdOpCode::WAKEUP_CORE));
}

void KDTU::suspend(KVPE &vpe) {
    // invalidate TLB and cache
    do_ext_cmd(vpe, static_cast<DTU::reg_t>(DTU::ExtCmdOpCode::INV_TLB));
    do_ext_cmd(vpe, static_cast<DTU::reg_t>(DTU::ExtCmdOpCode::INV_CACHE));

    // disable paging
    alignas(DTU_PKG_SIZE) DTU::reg_t status = 0;
    Sync::compiler_barrier();
    write_mem(vpe, DTU::dtu_reg_addr(DTU::DtuRegs::STATUS), &status, sizeof(status));
}

void KDTU::injectIRQ(KVPE &vpe) {
    do_ext_cmd(vpe, static_cast<DTU::reg_t>(DTU::ExtCmdOpCode::INJECT_IRQ) | (0x40 << 2));
}

void KDTU::config_pf_remote(KVPE &vpe, int ep) {
    static_assert(static_cast<int>(DTU::DtuRegs::STATUS) == 0, "STATUS wrong");
    static_assert(static_cast<int>(DTU::DtuRegs::ROOT_PT) == 1, "ROOT_PT wrong");
    static_assert(static_cast<int>(DTU::DtuRegs::PF_EP) == 2, "PF_EP wrong");

    uint64_t rootpt = vpe.address_space()->root_pt();
    if(!rootpt) {
        // TODO read the root pt from the core; the HW sets it atm for apps that are started at boot
        uintptr_t addr = DTU::dtu_reg_addr(DTU::DtuRegs::ROOT_PT);
        read_mem_at(vpe.core(), vpe.id(), addr, &rootpt, sizeof(rootpt));
    }
    else {
        clear_pt(rootpt);

        // insert recursive entry
        uintptr_t addr = DTU::noc_to_virt(rootpt);
        DTU::pte_t pte = rootpt | DTU::PTE_RWX;
        write_mem_at(MEMORY_CORE, 0, addr + PAGE_SIZE - sizeof(pte), &pte, sizeof(pte));
    }

    alignas(DTU_PKG_SIZE) DTU::reg_t dtuRegs[3];
    dtuRegs[static_cast<size_t>(DTU::DtuRegs::STATUS)]  = DTU::StatusFlags::PAGEFAULTS;
    dtuRegs[static_cast<size_t>(DTU::DtuRegs::ROOT_PT)] = rootpt;
    dtuRegs[static_cast<size_t>(DTU::DtuRegs::PF_EP)]   = ep;
    Sync::compiler_barrier();
    write_mem(vpe, DTU::dtu_reg_addr(DTU::DtuRegs::STATUS), dtuRegs, sizeof(dtuRegs));
}

static uintptr_t get_pte_addr(uintptr_t virt, int level) {
    static uintptr_t recMask = (DTU::PTE_REC_IDX << (PAGE_BITS + DTU::LEVEL_BITS * 2)) |
                               (DTU::PTE_REC_IDX << (PAGE_BITS + DTU::LEVEL_BITS * 1)) |
                               (DTU::PTE_REC_IDX << (PAGE_BITS + DTU::LEVEL_BITS * 0));

    // at first, just shift it accordingly.
    virt >>= PAGE_BITS + level * DTU::LEVEL_BITS;
    virt <<= DTU::PTE_BITS;

    // now put in one PTE_REC_IDX's for each loop that we need to take
    int shift = level + 1;
    uintptr_t remMask = (1UL << (PAGE_BITS + DTU::LEVEL_BITS * (DTU::LEVEL_CNT - shift))) - 1;
    virt |= recMask & ~remMask;

    // finally, make sure that we stay within the bounds for virtual addresses
    // this is because of recMask, that might actually have too many of those.
    virt &= (1UL << (DTU::LEVEL_CNT * DTU::LEVEL_BITS + PAGE_BITS)) - 1;
    return virt;
}

void KDTU::map_page(KVPE &vpe, uintptr_t virt, uintptr_t phys, int perm) {
    // configure the memory EP once and use it for all accesses
    config_mem_local(_ep, vpe.core(), vpe.id(), 0, 0xFFFFFFFFFFFFFFFF);
    for(int level = DTU::LEVEL_CNT - 1; level >= 0; --level) {
        uintptr_t pteAddr = get_pte_addr(virt, level);
        DTU::pte_t pte;
        if(level > 0) {
            DTU::get().read(_ep, &pte, sizeof(pte), pteAddr);

            // create the pagetable on demand
            if(pte == 0) {
                // if we don't have a pagetable for that yet, unmapping is a noop
                if(perm == 0)
                    return;

                // TODO this is prelimilary
                uintptr_t addr = MainMemory::get().map().allocate(PAGE_SIZE);
                pte = DTU::build_noc_addr(MEMORY_CORE, addr) | DTU::PTE_RWX;
                LOG(PTES, "PE" << vpe.core() << ": lvl 1 PTE for "
                    << fmt(virt, "p") << ": " << fmt(pte, "#0x", 16));
                DTU::get().write(_ep, &pte, sizeof(pte), pteAddr);
            }

            assert((pte & DTU::PTE_IRWX) == DTU::PTE_RWX);
        }
        else {
            pte = phys | perm | DTU::PTE_I;
            LOG(PTES, "PE" << vpe.core() << ": lvl 0 PTE for "
                << fmt(virt, "p") << ": " << fmt(pte, "#0x", 16));
            DTU::get().write(_ep, &pte, sizeof(pte), pteAddr);
        }
    }
}

void KDTU::unmap_page(KVPE &vpe, uintptr_t virt) {
    map_page(vpe, virt, 0, 0);

    // TODO remove pagetables on demand

    // invalidate TLB entry
    do_ext_cmd(vpe, static_cast<DTU::reg_t>(DTU::ExtCmdOpCode::INV_PAGE) | (virt << 2));
}

void KDTU::invalidate_ep(KVPE &vpe, int ep) {
    alignas(DTU_PKG_SIZE) DTU::reg_t e[DTU::EP_REGS];
    memset(&e, 0, sizeof(e));
    Sync::compiler_barrier();
    write_mem(vpe, DTU::ep_regs_addr(ep), &e, sizeof(e));
}

void KDTU::invalidate_eps(KVPE &vpe) {
    DTU::reg_t *eps = new DTU::reg_t[DTU::EP_REGS * EP_COUNT];
    size_t total = sizeof(*eps) * DTU::EP_REGS * EP_COUNT;
    memset(eps, 0, total);
    Sync::compiler_barrier();
    write_mem(vpe, DTU::ep_regs_addr(0), eps, total);
    delete[] eps;
}

void KDTU::config_recv(void *e, uintptr_t buf, uint order, uint msgorder, int) {
    DTU::reg_t *ep = reinterpret_cast<DTU::reg_t*>(e);
    DTU::reg_t bufSize = static_cast<DTU::reg_t>(1) << (order - msgorder);
    DTU::reg_t msgSize = static_cast<DTU::reg_t>(1) << msgorder;
    ep[0] = (static_cast<DTU::reg_t>(DTU::EpType::RECEIVE) << 61) |
            ((msgSize & 0xFFFF) << 32) | ((bufSize & 0xFFFF) << 16) | 0;
    ep[1] = buf;
    ep[2] = 0;
}

void KDTU::config_recv_local(int ep, uintptr_t buf, uint order, uint msgorder, int flags) {
    config_recv(reinterpret_cast<DTU::reg_t*>(DTU::ep_regs_addr(ep)),
        buf, order, msgorder, flags);
}

void KDTU::config_recv_remote(KVPE &vpe, int ep, uintptr_t buf, uint order, uint msgorder, int flags,
        bool valid) {
    alignas(DTU_PKG_SIZE) DTU::reg_t e[DTU::EP_REGS];
    memset(&e, 0, sizeof(e));

    if(valid)
        config_recv(&e, buf, order, msgorder, flags);

    // write to PE
    Sync::compiler_barrier();
    write_mem(vpe, DTU::ep_regs_addr(ep), &e, sizeof(e));
}

void KDTU::config_send(void *e, label_t label, int dstcore, int dstvpe, int dstep,
        size_t msgsize, word_t) {
    DTU::reg_t *ep = reinterpret_cast<DTU::reg_t*>(e);
    ep[0] = (static_cast<DTU::reg_t>(DTU::EpType::SEND) << 61) |
            ((dstvpe & 0xFFFF) << 16) | (msgsize & 0xFFFF);
    // TODO hand out "unlimited" credits atm
    ep[1] = ((dstcore & 0xFF) << 24) | ((dstep & 0xFF) << 16) | 0xFFFF;
    ep[2] = label;
}

void KDTU::config_send_local(int ep, label_t label, int dstcore, int dstvpe, int dstep,
        size_t msgsize, word_t credits) {
    config_send(reinterpret_cast<DTU::reg_t*>(DTU::ep_regs_addr(ep)),
        label, dstcore, dstvpe, dstep, msgsize, credits);
}

void KDTU::config_send_remote(KVPE &vpe, int ep, label_t label, int dstcore, int dstvpe, int dstep,
        size_t msgsize, word_t credits) {
    alignas(DTU_PKG_SIZE) DTU::reg_t e[DTU::EP_REGS];
    memset(&e, 0, sizeof(e));
    config_send(&e, label, dstcore, dstvpe, dstep, msgsize, credits);

    // write to PE
    Sync::compiler_barrier();
    write_mem(vpe, DTU::ep_regs_addr(ep), &e, sizeof(e));
}

void KDTU::config_mem(void *e, int dstcore, int dstvpe, uintptr_t addr, size_t size, int perm) {
    DTU::reg_t *ep = reinterpret_cast<DTU::reg_t*>(e);
    ep[0] = (static_cast<DTU::reg_t>(DTU::EpType::MEMORY) << 61) | (size & 0x1FFFFFFFFFFFFFFF);
    ep[1] = addr;
    ep[2] = ((dstvpe & 0xFFFF) << 12) | ((dstcore & 0xFF) << 4) | (perm & 0x7);
}

void KDTU::config_mem_local(int ep, int dstcore, int dstvpe, uintptr_t addr, size_t size) {
    config_mem(reinterpret_cast<DTU::reg_t*>(DTU::ep_regs_addr(ep)),
        dstcore, dstvpe, addr, size, DTU::R | DTU::W);
}

void KDTU::config_mem_remote(KVPE &vpe, int ep, int dstcore, int dstvpe, uintptr_t addr, size_t size, int perm) {
    alignas(DTU_PKG_SIZE) DTU::reg_t e[DTU::EP_REGS];
    memset(&e, 0, sizeof(e));
    config_mem(&e, dstcore, dstvpe, addr, size, perm);

    // write to PE
    Sync::compiler_barrier();
    write_mem(vpe, DTU::ep_regs_addr(ep), &e, sizeof(e));
}

void KDTU::reply_to(KVPE &vpe, int ep, int, word_t, label_t label, const void *msg, size_t size) {
    config_send_local(_ep, label, vpe.core(), vpe.id(), ep, size + DTU::HEADER_SIZE, size + DTU::HEADER_SIZE);
    DTU::get().send(_ep, msg, size, 0, 0);
}

void KDTU::write_mem(KVPE &vpe, uintptr_t addr, const void *data, size_t size) {
    write_mem_at(vpe.core(), vpe.id(), addr, data, size);
}

void KDTU::write_mem_at(int core, int vpe, uintptr_t addr, const void *data, size_t size) {
    config_mem_local(_ep, core, vpe, addr, size);
    DTU::get().write(_ep, data, size, 0);
}

void KDTU::read_mem_at(int core, int vpe, uintptr_t addr, void *data, size_t size) {
    config_mem_local(_ep, core, vpe, addr, size);
    DTU::get().read(_ep, data, size, 0);
}

}

File: src/apps/rust/kernel/src/arch/gem5/kdtu.rs
use base::cell::StaticCell;
use base::cfg;
use base::col::Vec;
use base::dtu::*;
use base::errors::{Code, Error};
use base::goff;
use base::kif::Perm;
use base::libc;
use base::util;
use core::intrinsics;

use arch::platform;
use pes::{INVALID_VPE, VPEId, VPEDesc};
use pes::rctmux;
use pes::vpemng;

pub struct State {
    dtu: [Reg; DTU_REGS],
    cmd: [Reg; CMD_REGS],
    eps: [Reg; EP_COUNT * EP_REGS],
    header: [ReplyHeader; HEADER_COUNT],
    header_off: usize,
}

impl State {
    pub fn new() -> State {
        State {
            dtu: [0; DTU_REGS],
            cmd: [0; CMD_REGS],
            eps: [0; EP_COUNT * EP_REGS],
            header: [ReplyHeader::default(); HEADER_COUNT],
            header_off: 0,
        }
    }

    pub fn set_dtu_reg(&mut self, reg: DtuReg, val: Reg) {
        self.dtu[reg.val as usize] = val;
    }

    pub fn get_ep(&self, ep: EpId) -> &[Reg] {
        &self.eps[ep * EP_REGS..(ep + 1) * EP_REGS]
    }
    pub fn get_ep_mut(&mut self, ep: EpId) -> &mut [Reg] {
        &mut self.eps[ep * EP_REGS..(ep + 1) * EP_REGS]
    }

    pub fn config_send(&mut self, ep: EpId, lbl: Label, pe: PEId, vpe: VPEId,
                       dst_ep: EpId, msg_size: usize, credits: u64) {
        let regs: &mut [Reg] = self.get_ep_mut(ep);
        regs[0] = (EpType::SEND.val << 61) |
                  ((vpe as Reg & 0xFFFF) << 16) |
                  (msg_size as Reg & 0xFFFF);
        regs[1] = ((pe as Reg & 0xFF) << 40) |
                  ((dst_ep as Reg & 0xFF) << 32) |
                  (credits << 16) |
                  (credits << 0);
        regs[2] = lbl;
    }

    pub fn config_recv(&mut self, ep: EpId, buf: goff, ord: i32, msg_ord: i32, header: usize) {
        let regs: &mut [Reg] = self.get_ep_mut(ep);
        let buf_size = 1 << (ord - msg_ord);
        let msg_size = 1 << msg_ord;
        regs[0] = (EpType::RECEIVE.val << 61) |
                  ((msg_size as Reg & 0xFFFF) << 32) |
                  ((buf_size as Reg & 0xFFFF) << 16) |
                  ((header as Reg) << 5);
        regs[1] = buf as Reg;
        regs[2] = 0;
    }

    pub fn config_mem(&mut self, ep: EpId, pe: PEId, vpe: VPEId, addr: goff, size: usize, perm: Perm) {
        let regs: &mut [Reg] = self.get_ep_mut(ep);
        regs[0] = (EpType::MEMORY.val << 61) |
                  (size as Reg & 0x1FFF_FFFF_FFFF_FFFF);
        regs[1] = addr as Reg;
        regs[2] = ((vpe as Reg & 0xFFFF) << 12) |
                  ((pe as Reg & 0xFF) << 4) |
                  (perm.bits() as Reg & 0x7);
    }

    pub fn invalidate(&mut self, ep: EpId, check: bool) -> Result<(), Error> {
        let regs: &mut [Reg] = self.get_ep_mut(ep);
        if check && (regs[0] >> 61) == EpType::SEND.val {
            if (regs[1] >> 16) & 0xFFFF != (regs[1] & 0xFFFF) {
                return Err(Error::new(Code::InvArgs));
            }
        }

        regs[0] = 0;
        regs[1] = 0;
        regs[2] = 0;
        Ok(())
    }

    pub fn restore(&mut self, vpe: &VPEDesc, vpe_id: VPEId) {
        // TODO copy the receive buffers back to the SPM

        // TODO reenable pagefaults
        self.set_dtu_reg(DtuReg::STATUS, 0);
        self.set_dtu_reg(DtuReg::VPE_ID, vpe_id as Reg);
        self.set_dtu_reg(DtuReg::IDLE_TIME, 0);

        // restore registers
        let size = util::size_of_val(&self.dtu) +
                   util::size_of_val(&self.cmd) +
                   util::size_of_val(&self.eps);
        KDTU::get().try_write_mem(vpe, BASE_ADDR as goff,
            self.dtu.as_mut_ptr() as *mut u8, size).unwrap();

        // restore headers (we've already set the VPE id)
        let hdr_size = util::size_of_val(&self.header);
        KDTU::get().state.config_mem(KTMP_EP, vpe.pe_id(), vpe_id,
            (BASE_ADDR + size) as goff, hdr_size, Perm::W);
        KDTU::get().write_ep_local(KTMP_EP);

        DTU::write(KTMP_EP, self.header.as_mut_ptr() as *mut u8, hdr_size, 0, CmdFlags::NOPF).unwrap();
    }

    pub fn reset(&mut self) {
        self.set_dtu_reg(DtuReg::EXT_CMD, ExtCmdOpCode::RESET.val as Reg);
    }
}

pub struct KDTU {
    state: State,
    buf: Vec<u8>,
}

pub const KSYS_EP: EpId   = 0;
pub const KSRV_EP: EpId   = 1;
pub const KTMP_EP: EpId   = 2;

static INST: StaticCell<Option<KDTU>> = StaticCell::new(None);

impl KDTU {
    pub fn init() {
        INST.set(Some(KDTU {
            state: State::new(),
            buf: vec![0u8; 4096],
        }));

        // set our own VPE id
        Self::get().do_set_vpe_id(
            &VPEDesc::new_kernel(INVALID_VPE),
            vpemng::KERNEL_VPE
        ).unwrap();
    }

    pub fn get() -> &'static mut KDTU {
        INST.get_mut().as_mut().unwrap()
    }

    fn do_set_vpe_id(&mut self, vpe: &VPEDesc, nid: VPEId) -> Result<(), Error> {
        let vpe_id = nid as Reg;
        self.try_write_slice(vpe, DTU::dtu_reg_addr(DtuReg::VPE_ID) as goff, &[vpe_id])
    }

    pub fn write_ep_local(&mut self, ep: EpId) {
        DTU::set_ep(ep, self.state.get_ep(ep));
    }

    pub fn invalidate_ep_remote(&mut self, vpe: &VPEDesc, ep: EpId) -> Result<(), Error> {
        let reg = ExtCmdOpCode::INV_EP.val | (ep << 3) as Reg;
        self.try_write_slice(vpe, DTU::dtu_reg_addr(DtuReg::EXT_CMD) as goff, &[reg])
    }

    pub fn read_obj<T>(&mut self, vpe: &VPEDesc, addr: goff) -> T {
        self.try_read_obj(vpe, addr).unwrap()
    }

    pub fn try_read_obj<T>(&mut self, vpe: &VPEDesc, addr: goff) -> Result<T, Error> {
        let mut obj: T = unsafe { intrinsics::uninit() };
        self.try_read_mem(vpe, addr, &mut obj as *mut T as *mut u8, util::size_of::<T>())?;
        Ok(obj)
    }

    pub fn read_mem(&mut self, vpe: &VPEDesc, addr: goff, data: *mut u8, size: usize) {
        assert!(vpe.vpe().is_none());
        self.try_read_mem(vpe, addr, data, size).unwrap();
    }

    pub fn try_read_mem(&mut self, vpe: &VPEDesc, addr: goff, data: *mut u8, size: usize) -> Result<(), Error> {
        self.state.config_mem(KTMP_EP, vpe.pe_id(), vpe.vpe_id(), addr, size, Perm::R);
        self.write_ep_local(KTMP_EP);

        DTU::read(KTMP_EP, data, size, 0, CmdFlags::NOPF)
    }

    pub fn write_slice<T>(&mut self, vpe: &VPEDesc, addr: goff, sl: &[T]) {
        self.write_mem(vpe, addr, sl.as_ptr() as *const u8, sl.len() * util::size_of::<T>());
    }

    pub fn try_write_slice<T>(&mut self, vpe: &VPEDesc, addr: goff, sl: &[T]) -> Result<(), Error> {
        self.try_write_mem(vpe, addr, sl.as_ptr() as *const u8, sl.len() * util::size_of::<T>())
    }

    pub fn write_mem(&mut self, vpe: &VPEDesc, addr: goff, data: *const u8, size: usize) {
        assert!(vpe.vpe().is_none());
        self.try_write_mem(vpe, addr, data, size).unwrap();
    }

    pub fn try_write_mem(&mut self, vpe: &VPEDesc, addr: goff, data: *const u8, size: usize) -> Result<(), Error> {
        self.state.config_mem(KTMP_EP, vpe.pe_id(), vpe.vpe_id(), addr, size, Perm::W);
        self.write_ep_local(KTMP_EP);

        DTU::write(KTMP_EP, data, size, 0, CmdFlags::NOPF)
    }

    pub fn send_to(&mut self, vpe: &VPEDesc, ep: EpId, lbl: Label, msg: *const u8, size: usize,
                   rpl_lbl: Label, rpl_ep: EpId) -> Result<(), Error> {
        let msg_size = size + util::size_of::<Header>();
        self.state.config_send(KTMP_EP, lbl, vpe.pe_id(), vpe.vpe_id(), ep, msg_size, CREDITS_UNLIM);
        self.write_ep_local(KTMP_EP);

        let sender = (platform::kernel_pe() as Reg) |
                     ((vpemng::KERNEL_VPE as Reg) << 8) |
                     ((EP_COUNT as Reg) << 24) |
                     ((rpl_ep as Reg) << 32);

        DTU::send_to(KTMP_EP, msg, size, rpl_lbl, rpl_ep, sender as u64)
    }

    pub fn clear(&mut self, dst_vpe: &VPEDesc, mut dst_addr: goff, size: usize) -> Result<(), Error> {
        unsafe {
            libc::memset(self.buf.as_ptr() as *mut libc::c_void, 0, self.buf.len());
        }

        let mut rem = size;
        while rem > 0 {
            let amount = util::min(rem, self.buf.len());
            let buf = self.buf.as_ptr();
            self.try_write_mem(dst_vpe, dst_addr, buf, amount)?;
            dst_addr += amount as goff;
            rem -= amount;
        }
        Ok(())
    }

    pub fn copy(&mut self, dst_vpe: &VPEDesc, mut dst_addr: goff,
                           src_vpe: &VPEDesc, mut src_addr: goff,
                           size: usize) -> Result<(), Error> {
        let mut rem = size;
        while rem > 0 {
            let amount = util::min(rem, self.buf.len());
            let buf = self.buf.as_mut_ptr();
            self.try_read_mem(src_vpe, src_addr, buf, amount)?;
            self.try_write_mem(dst_vpe, dst_addr, buf, amount)?;
            src_addr += amount as goff;
            dst_addr += amount as goff;
            rem -= amount;
        }
        Ok(())
    }

    pub fn recv_msgs(&mut self, ep: EpId, buf: goff, ord: i32, msg_ord: i32) -> Result<(), Error> {
        if self.state.header_off + (1 << (ord - msg_ord)) > HEADER_COUNT {
            return Err(Error::new(Code::NoSpace));
        }

        let off = self.state.header_off;
        self.state.config_recv(ep, buf, ord, msg_ord, off);
        self.state.header_off += 1 << (ord - msg_ord);
        self.write_ep_local(ep);

        Ok(())
    }

    pub fn write_ep_remote(&mut self, vpe: &VPEDesc, ep: EpId, regs: &[Reg]) -> Result<(), Error> {
        self.try_write_slice(vpe, DTU::ep_regs_addr(ep) as goff, regs)
    }

    fn do_ext_cmd(&mut self, vpe: &VPEDesc, cmd: Reg) -> Result<(), Error> {
        self.try_write_slice(vpe, DTU::dtu_reg_addr(DtuReg::EXT_CMD) as goff, &[cmd])
    }

    pub fn invalidate_tlb(&mut self, vpe: &VPEDesc) -> Result<(), Error> {
        self.do_ext_cmd(vpe, ExtCmdOpCode::INV_TLB.val)
    }

    pub fn invlpg_remote(&mut self, vpe: &VPEDesc, mut virt: goff) -> Result<(), Error> {
        virt &= !cfg::PAGE_MASK as goff;
        let cmd = ExtCmdOpCode::INV_PAGE.val | (virt << 3) as u64;
        self.do_ext_cmd(vpe, cmd)
    }

    pub fn reset(&mut self, vpe: &VPEDesc) -> Result<(), Error> {
        // TODO temporary
        let id = INVALID_VPE as Reg;
        self.try_write_slice(vpe, DTU::dtu_reg_addr(DtuReg::VPE_ID) as goff, &[id])
    }

    pub fn wakeup(&mut self, vpe: &VPEDesc, addr: goff) -> Result<(), Error> {
        let cmd = ExtCmdOpCode::WAKEUP_CORE.val | (addr << 3) as u64;
        self.do_ext_cmd(vpe, cmd)
    }

    pub fn inject_irq(&mut self, vpe: &VPEDesc, cmd: Reg) -> Result<(), Error> {
        self.try_write_slice(vpe, DTU::dtu_req_addr(ReqReg::EXT_REQ) as goff, &[cmd])
    }

    pub fn write_swstate(&mut self, vpe: &VPEDesc, flags: u64, notify: u64) -> Result<(), Error> {
        let vals = [notify, flags];
        self.try_write_slice(vpe, rctmux::YIELD_ADDR, &vals)
    }

    pub fn write_swflags(&mut self, vpe: &VPEDesc, flags: u64) -> Result<(), Error> {
        self.try_write_slice(vpe, rctmux::FLAGS_ADDR, &[flags])
    }

    pub fn read_swflags(&mut self, vpe: &VPEDesc) -> Result<u64, Error> {
        self.try_read_obj(vpe, rctmux::FLAGS_ADDR)
    }
}
